### PR TITLE
Add new ceremony registration system

### PR DESCRIPTION
**Pull request by TheShadowEevee**
WIP

This PR is an overall update to the passport ceremony system, per requirements from @MatthewStanciu. Contributions are welcome on the [Ceremony-Registration-Enhancements branch](https://github.com/purduehackers/passport-issuing-office/tree/Ceremony-Registration-Enhancements).

Will update the following as I remember to update the following
Task list open to comments / corrections
| No. | Task | Priority |
|-----|------|----------|
|  1  |  Create Dropdown of Passport Ceremonies | 0 |
|  2  |  Admin Page | 1 |
|  3  |  Make clear to users what they are signing up for | 1 |
|  4  |  PDF Generation | 3 |
|  5  |  Group Passports by Ceremony Date in App | 5 |
|  6  |  Discord Bot | 5 |

### Task 1 (Complete - Pending Review)
- [X] Create Form Field for Date submission
- [X] Post Field on Generate
- [X] Set to ~~Null~~, ~~Undefined~~, unique string when not registering
- [X] Make Field a Dropdown
- [X] Get List from List of Ceremonies
  - [X] New DB Table of Ceremonies?
  - [X] Change database's sample data date format (to YYYY-MM-DDTHH:MM:SS.XXXXZ), where X is whatever the letter for milliseconds is
- [X] Make dropdown adopt the selected date as its label

### Task 2 (Blocked - Auth.js limited to one session)
- [x] Create an Admin Page for those with Admin Passports
- [ ] Log in using Passport (admin scope)
  - [ ] Figure out how to do Passport and Discord Auth at the same time
- [x] Create Ceremonies
- [x] Delete Ceremonies
- [x] Change Ceremony Date from Passport ID or Discord ID
- [x] Add link to playground
  - [x] Check if logged in Discord ID matches a user with admin role?

### Task 3 (Mostly Complete - Pending Review)
- [X] Create a Dialog when attempting to register for a ceremony
- [ ] Add some text explaining the ceremonies
- [X] Replace the generate button when registering
- [X] Add a Button in the dialog to generate the passport
- [X] Make dialog close on generate
- [x] Button is smaller. Why??
- [X] Fix dialog box size
- [ ] Other Cleanup
- [x] Change Page Layout

### Task 4
- [ ] TBD

### Task 5 - See [passport-authority](https://github.com/purduehackers/passport-authority)
- [ ] Group Passports by their Ceremony in Passport App 

### Task 6
- [ ] DM on Passport Creation with verification of ceremony date, image of passport?
- [ ] DM 7p day before the ceremony

### CODE REVIEW

**Comment by TheShadowEevee** [comment]
Force reverted to commit fd3098c "Pain".

Reverted commit history is available here:
https://github.com/purduehackers/passport-issuing-office/commits/9f1c7a2285ce577bcb1328a471302bdb97aed6d1/

File: lib/actions.ts
Comment (MatthewStanciu) [inline lib/actions.ts:105]: Idk if stupid question, but why are these async?

File: prisma/schema.prisma
Comment (MatthewStanciu) [inline prisma/schema.prisma:43]: sanity check, is this reflecting a migration previously applied in the id repo?

File: lib/actions.ts
Comment (TheShadowEevee) [reply lib/actions.ts:105]: They fit best in that file, but "Server actions must be async functions" so they were async tagged

Comment (MatthewStanciu) [reply lib/actions.ts:105]: lol TIL

File: prisma/schema.prisma
Comment (TheShadowEevee) [reply prisma/schema.prisma:43]: My comment on this vanished :)
(No, but is now)

**Comment by TheShadowEevee** [comment]
Should probably replace the placeholder text either before or right after merging

File: components/admin.tsx
Comment (MatthewStanciu) [inline components/admin.tsx:512]: should prob be removed lol

File: components/playground.tsx
Comment (MatthewStanciu) [inline components/playground.tsx:571]: Screenshot 2024-09-24 at 8 02 39 PM

I still think this looks weird. I suggest removing the padding between the heading and the text.

Also, let's change placeholder text to:

"By clicking "Register My Passport", you're signing up to make your own passport at the passport ceremony on <date> at <time>. Please show up on time so that you have enough time to make your passport. If you're late, we will start without you & you will need to re-register for the next one.

"We spend real money and human labor on each passport, so please make sure you're there! If plans change & you can't make it, please post in #lounge or DM @matthew and let us know so that we don't prep the materials for your passport."

**Review by MatthewStanciu — COMMENTED** [review]
Couple things

File: components/admin.tsx
Comment (TheShadowEevee) [reply components/admin.tsx:522]: Forgot about this lmao

**Comment by TheShadowEevee** [comment]
Broke a thing, will fix soon

File: components/playground.tsx
Comment (MatthewStanciu) [inline components/playground.tsx:280]: Let's do:

"Register for an hour-long passport ceremony to turn your data page into a real-life passport."

File: app/page.tsx
Comment (TheShadowEevee) [reply app/page.tsx:104]: I can remove this, but it's set to only display if the instance isn't a production environment; it won't appear once merged

File: components/playground.tsx
Comment (TheShadowEevee) [reply components/playground.tsx]: Resolved in [00d29d9](https://github.com/purduehackers/passport-issuing-office/pull/52/commits/00d29d972990892c27894b10c9ce4850b3c19ed8)

File: components/admin.tsx
Comment (TheShadowEevee) [reply components/admin.tsx]: Resolved in [00d29d9](https://github.com/purduehackers/passport-issuing-office/pull/52/commits/00d29d972990892c27894b10c9ce4850b3c19ed8)

File: components/playground.tsx
Comment (TheShadowEevee) [reply components/playground.tsx]: Resolved in [00d29d9](https://github.com/purduehackers/passport-issuing-office/pull/52/commits/00d29d972990892c27894b10c9ce4850b3c19ed8)

File: components/playground.tsx
Comment (TheShadowEevee) [reply components/playground.tsx:408]: Resolved in [00d29d9](https://github.com/purduehackers/passport-issuing-office/pull/52/commits/00d29d972990892c27894b10c9ce4850b3c19ed8)

File: components/playground.tsx
Comment (TheShadowEevee) [reply components/playground.tsx:330]: Resolved in [0281423](https://github.com/purduehackers/passport-issuing-office/pull/52/commits/028142374fd792ac613a387073508a59e35c6fe5)

File: app/page.tsx
Comment (MatthewStanciu) [reply app/page.tsx:99]: Oops, thought that was hardcoded

File: components/playground.tsx
Comment (TheShadowEevee) [reply components/playground.tsx]: Resolved in [f122838](https://github.com/purduehackers/passport-issuing-office/pull/52/commits/f122838cfd8f85d9ac4a4e424a901a1a1e111c1e)

File: components/playground.tsx
Comment (MatthewStanciu) [inline components/playground.tsx:311]: Screenshot 2024-09-24 at 9 03 54 PM

Nit, non-blocking: it just occurred to me that this is weird. It should be a single radio, or a button that you can toggle, or something, instead of a "yes" or "no" radio. There's no reason to check "no" except maybe to undo a "yes".

File: components/playground.tsx
Comment (MatthewStanciu) [inline components/playground.tsx:564]: Screenshot 2024-09-24 at 9 06 32 PM

Idk I still feel like this needs less of a gap between the heading and description, like one or two lines less

File: components/playground.tsx
Comment (TheShadowEevee) [reply components/playground.tsx]: Resolved in [3480db5](https://github.com/purduehackers/passport-issuing-office/pull/52/commits/3480db5711f75ef911b6a079e0d9deb8036ef6a8), if the switch is acceptable

Comment (TheShadowEevee) [reply components/playground.tsx]: image
image

File: components/playground.tsx
Comment (TheShadowEevee) [reply components/playground.tsx:549]: Shortened a bit more in [4eed739](https://github.com/purduehackers/passport-issuing-office/pull/52/commits/4eed7391fb891304bafd230079d42a0cb532328c), see if that's better

Comment (MatthewStanciu) [reply components/playground.tsx:549]: Screenshot 2024-09-24 at 9 57 35 PM

Looks great now

File: components/playground.tsx
Comment (MatthewStanciu) [reply components/playground.tsx]: Switch is a great idea, but it's hard to see when it's in an off state

File: app/page.tsx
Comment (MatthewStanciu) [inline app/page.tsx:80]: Let's do:

You're registered for the passport ceremony happening on <date> at <time>!

Keep it amber

Comment (MatthewStanciu) [reply app/page.tsx:80]: Also make sure it doesn't show if the ceremony is in the past or the passport is activated

Comment (TheShadowEevee) [reply app/page.tsx]: Resolved in [0dd1bd9](https://github.com/purduehackers/passport-issuing-office/pull/52/commits/0dd1bd90777c2dde7ee3c25d4d121a904f74843b)

File: app/api/generate-full-frame/route.ts
Comment (MatthewStanciu) [inline app/api/generate-full-frame/route.ts:33]: nit:

```suggestion
				sendToDb,
```

File: app/page.tsx
Comment (MatthewStanciu) [inline app/page.tsx:77]:
```suggestion
							<p>
								You&apos;re registered for the passport ceremony happening on {getCeremonyTimeStringDate(latestPassport.ceremony_time)} at {getCeremonyTimeStringTime(latestPassport.ceremony_time)}!
							</p>
							<p>
								If you need to cancel, please let us know in #lounge, or re-register for a different time.
							</p>
```

File: components/admin.tsx
Comment (MatthewStanciu) [inline components/admin.tsx:225]:
```suggestion
export default function AdminPage() {
```

File: components/admin.tsx
Comment (MatthewStanciu) [inline components/admin.tsx:744]: any reason why these are snake_case? feels out of place

File: components/playground.tsx
Comment (MatthewStanciu) [inline components/playground.tsx:152]: love it

File: lib/ceremony-data.tsx
Comment (MatthewStanciu) [inline lib/ceremony-data.tsx:20]: TIL

File: lib/ceremony-dropdown.tsx
Comment (MatthewStanciu) [inline lib/ceremony-dropdown.tsx:26]: If the goal is to fetch this data when the page loads, you should instead use `swr` or fetch the data in a server component and pass it down to the client, IMO. Data fetching in useEffect works but it's not very performant

File: lib/ceremony-dropdown.tsx
Comment (MatthewStanciu) [inline lib/ceremony-dropdown.tsx:39]: it feels weird that this value is snake_case, i'd rather it be camelCase for consistency with other object properties in this repo

File: lib/parse-form-data.ts
Comment (MatthewStanciu) [inline lib/parse-form-data.ts:32]: huh why's this `as unknown`? that's usually a sign that something's going wrong with types

**Review by MatthewStanciu — APPROVED** [review]
couple nits, but nothing that should block this pr. thank you so much for your work here 🙏 💛

File: lib/parse-form-data.ts
Comment (TheShadowEevee) [reply lib/parse-form-data.ts:32]: Forgot this was there, I needed to clear this error
image

No, I cannot find any reason that's happening. It's always a boolean and always acts as such. Agh.

File: lib/ceremony-dropdown.tsx
Comment (TheShadowEevee) [reply lib/ceremony-dropdown.tsx:39]: The reason here is cause I'm pulling a value directly from the database, which uses snake_case (see auth.ts for another example)

File: lib/parse-form-data.ts
Comment (MatthewStanciu) [reply lib/parse-form-data.ts:32]: whatever, we can figure it out later

File: components/admin.tsx
Comment (TheShadowEevee) [reply components/admin.tsx]: Resolved in [b9bfab8](https://github.com/purduehackers/passport-issuing-office/pull/52/commits/b9bfab814516ac0d795573f35a8422e1878bf802)

**Pull request merged**
